src/TiledArray/math/scalapack/chol.h

```diff
@@ -0,0 +1,306 @@
+/*
+ *  This file is a part of TiledArray.
+ *  Copyright (C) 2020 Virginia Tech
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  David Williams-Young
+ *  Computational Research Division, Lawrence Berkeley National Laboratory
+ *
+ *  chol.h
+ *  Created:    8 June, 2020
+ *
+ */
+#ifndef TILEDARRAY_MATH_SCALAPACK_CHOL_H__INCLUDED
+#define TILEDARRAY_MATH_SCALAPACK_CHOL_H__INCLUDED
+
+#include <TiledArray/config.h>
+#if TILEDARRAY_HAS_SCALAPACK
+
+#include <TiledArray/conversions/block_cyclic.h>
+#include <TiledArray/math/scalapack/util.h>
+
+#include <scalapackpp/factorizations/potrf.hpp>
+#include <scalapackpp/matrix_inverse/trtri.hpp>
+#include <scalapackpp/linear_systems/posv.hpp>
+#include <scalapackpp/linear_systems/trtrs.hpp>
+
+namespace TiledArray {
+
+/**
+ *  @brief Compute the Cholesky factorization of a HPD rank-2 tensor
+ *
+ *  A(i,j) = L(i,k) * conj(L(j,k))
+ *
+ *  Example Usage:
+ *
+ *  auto L = cholesky(A, ...)
+ *
+ *  @tparam Array Input array type, must be convertable to BlockCyclicMatrix
+ *
+ *  @param[in] A           Input array to be diagonalized. Must be rank-2
+ *  @param[in] NB          ScaLAPACK blocking factor. Defaults to 128
+ *  @param[in] l_trange    TiledRange for resulting Cholesky factor. If left empty,
+ *                         will default to array.trange()
+ *
+ *  @returns The lower triangular Cholesky factor L in TA format
+ */
+template <typename Array>
+auto cholesky( const Array& A, size_t NB = 128, TiledRange l_trange = TiledRange() ) {
+
+  using value_type = typename Array::element_type;
+
+  auto& world = A.world();
+  auto world_comm = world.mpi.comm().Get_mpi_comm();
+  blacspp::Grid grid = blacspp::Grid::square_grid(world_comm);
+
+  world.gop.fence(); // stage ScaLAPACK execution
+  auto matrix = array_to_block_cyclic( A, grid, NB, NB );
+  world.gop.fence(); // stage ScaLAPACK execution
+
+  auto [M, N] = matrix.dims();
+  if( M != N )
+    throw std::runtime_error("Matrix must be square for Cholesky");
```

Comment:
should throw TiledArray::Exception (here and elsewhere)

Reply:
`std::runtime_error` -> `TA_EXCEPTION`